myia/pipeline.py

```diff
@@ -0,0 +1,516 @@
+
+    mode = 'override'
+
+
+class Reset(MergeMode):
+    """Throw away the previous value."""
+
+    mode = 'reset'
+
+
+_cleanup_map = TypeMap()
+
+
+@_cleanup_map.register(object)
+def _cleanup_object(value):
+    return value
+
+
+@_cleanup_map.register(MergeMode)
+def _cleanup_MergeMode(mm):
+    return mm.value
+
+
+@_cleanup_map.register(dict)
+def _cleanup_dict(d):
+    return type(d)({k: cleanup(v) for k, v in d.items() if v is not DELETE})
+
+
+def _cleanup_sequence(xs):
+    return type(xs)(cleanup(x) for x in xs)
+
+
+_cleanup_map.register(tuple, _cleanup_sequence)
+_cleanup_map.register(list, _cleanup_sequence)
+_cleanup_map.register(set, _cleanup_sequence)
+
+
+def cleanup(x):
+    """Remove all MergeMode and DELETE instances from the data."""
+    return _cleanup_map[type(x)](x)
+
+
+_merge_map = TypeMap(discover=lambda cls: getattr(cls, '__merge__', None))
+
+
+@_merge_map.register(dict)
+def _merge_dict(d1, d2, mode):
+    if mode == 'reset':
+        return type(d1)(d2)
+
+    rval = type(d1)()
+    for k, v in d1.items():
+        if k in d2:
+            v2 = d2[k]
+            if v2 is DELETE:
+                pass
+            else:
+                rval[k] = merge(v, v2, mode)
+        else:
+            rval[k] = v
+    for k, v in d2.items():
+        if k not in d1:
+            rval[k] = cleanup(v)
+    return rval
+
+
+@_merge_map.register(tuple)
+def _merge_tuple(xs, ys, mode):
+    xs = cleanup(xs)
+    ys = cleanup(ys)
+    if mode == 'merge':
+        return xs + ys
+    else:
+        return ys
+
+
+@_merge_map.register(list)
+def _merge_list(xs, ys, mode):
+    xs = cleanup(xs)
+    ys = cleanup(ys)
+    if mode == 'merge':
+        return xs + ys
+    else:
+        return ys
+
+
+@_merge_map.register(set)
+def _merge_set(xs, ys, mode):
+    xs = cleanup(xs)
+    ys = cleanup(ys)
+    if mode == 'merge':
+        return xs | ys
+    else:
+        return ys
+
+
+@_merge_map.register(object)
+def _merge_object(a, b, mode):
+    return cleanup(b)
+
+
+def merge(a, b, mode=MergeMode.mode):
+    """Merge two data structures.
+
+    Arguments:
+        a: The original data.
+        b: The data to merge.
+        mode:
+            'merge': (default) Sequences will be concatenated, sets merged,
+                and dictionaries merged according to their keys.
+            'override': Dictionaries are merged, but sequences are not
+                concatenated.
+            'reset': b is returned, or takes primacy over the data in a.
+    """
+    if isinstance(b, MergeMode):
+        mode = b.mode
+        b = b.value
+    assert not isinstance(a, MergeMode)
+    return _merge_map[type(a)](a, b, mode)
+
+
+def _filter_keywords(f, kw):
+    spec = inspect.getfullargspec(f)
+    if spec.varkw:
+        return kw, {}
+    valid = spec.args + spec.kwonlyargs
+
+    good = {k: v for k, v in kw.items() if k in valid}
+    bad = {k: v for k, v in kw.items() if k not in valid}
+
+    return good, bad
+
+
+class NS:
```

Comment:
This too.

Reply:
Yeah, the first half of `pipeline.py` should be in utils, at which point I think utils should probably be split in a few files. I wanted to do that as a separate PR afterwards.

Reply:
ok